File: src/main/java/com/milaboratory/mixcr/cli/postanalysis/CommandPaExportPlotsOverlap.java
/*
 * Copyright (c) 2014-2022, MiLaboratories Inc. All Rights Reserved
 *
 * Before downloading or accessing the software, please read carefully the
 * License Agreement available at:
 * https://github.com/milaboratory/mixcr/blob/develop/LICENSE
 *
 * By downloading or accessing the software, you accept and agree to be bound
 * by the terms of the License Agreement. If you do not want to agree to the terms
 * of the Licensing Agreement, you must not download or access the software.
 */
package com.milaboratory.mixcr.cli.postanalysis;

import com.milaboratory.miplots.Position;
import com.milaboratory.mixcr.basictypes.Clone;
import com.milaboratory.mixcr.postanalysis.PostanalysisResult;
import com.milaboratory.mixcr.postanalysis.overlap.OverlapType;
import com.milaboratory.mixcr.postanalysis.plots.*;
import com.milaboratory.mixcr.postanalysis.ui.CharacteristicGroup;
import com.milaboratory.mixcr.postanalysis.ui.PostanalysisParametersOverlap;
import jetbrains.letsPlot.intern.Plot;
import org.jetbrains.kotlinx.dataframe.DataFrame;
import picocli.CommandLine.Command;
import picocli.CommandLine.Option;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

@Command(name = "overlap",
        sortOptions = false,
        separator = " ",
        description = "Export overlap heatmaps")
public class CommandPaExportPlotsOverlap extends CommandPaExportPlotsHeatmapWithGroupBy {
    @Option(description = "Don't add dendrograms",
            names = {"--no-dendro"})
    public boolean noDendro;

    @Option(description = "Add color key layer; prefix 'x_' (add to the bottom) or 'y_' (add to the left) should be used.",
            names = {"--color-key"})
    public List<String> colorKey = new ArrayList<>();

    @Option(description = "Fill diagonal line",
            names = {"--fill-diagonal"}
    )
    public boolean fillDiagonal = false;

    @Option(description = "Select specific metrics to export.",
            names = {"--metric"})
    public List<String> metrics;

    DataFrame<OverlapRow> filterOverlap(DataFrame<OverlapRow> df) {
        if (filterByMetadata != null) {
            for (String f : filterByMetadata) {
                Filter filter = MetadataKt.parseFilter(metadata(), f);
                df = Overlap.INSTANCE.filterOverlap(filter, df);
            }
        }
        return df;
    }

    private List<OverlapType> metricsFilter() {
        if (metrics == null || metrics.isEmpty())
            return null;
        return metrics.stream()
                .map(OverlapType::byNameOrThrow)
                .collect(Collectors.toList());
    }

    @Override
    void run(PaResultByGroup result) {
        CharacteristicGroup<Clone, ?> ch = result.schema.getGroup(PostanalysisParametersOverlap.Overlap);
        PostanalysisResult paResult = result.result.forGroup(ch);
        DataFrame<?> metadata = metadata();

        List<OverlapType> metrics = metricsFilter();
        DataFrame<OverlapRow> df = Overlap.INSTANCE.dataFrame(
                paResult,
                metrics,
                fillDiagonal,
                metadata
        );
        df = filterOverlap(df);

        if (df.rowsCount() == 0)
            return;

        if (df.get("weight").distinct().toList().size() <= 1)
            return;

        List<String> colorKey = this.colorKey.stream()
                .map(it -> it.startsWith("x_") || it.startsWith("y_") ? it : "x_" + it)
                .collect(Collectors.toList());

        HeatmapParameters par = new HeatmapParameters(
                !noDendro,
                !noDendro,
                colorKey.stream()
                        .map(it -> new ColorKey(it, it.startsWith("x") ? Position.Bottom : Position.Left))
                        .collect(Collectors.toList()),
                groupBy,
                hLabelsSize,
                vLabelsSize,
                false,
                parsePallete(),
                width,
                height
        );

        List<Plot> plots = Overlap.INSTANCE.plots(df, par);
        writePlots(result.group, plots);
    }
}

File: src/main/kotlin/com/milaboratory/mixcr/cli/postanalysis/CommandPaExportPlotsOverlap.kt
/*
 * Copyright (c) 2014-2022, MiLaboratories Inc. All Rights Reserved
 *
 * Before downloading or accessing the software, please read carefully the
 * License Agreement available at:
 * https://github.com/milaboratory/mixcr/blob/develop/LICENSE
 *
 * By downloading or accessing the software, you accept and agree to be bound
 * by the terms of the License Agreement. If you do not want to agree to the terms
 * of the Licensing Agreement, you must not download or access the software.
 */
package com.milaboratory.mixcr.cli.postanalysis

import com.milaboratory.miplots.Position.Bottom
import com.milaboratory.miplots.Position.Left
import com.milaboratory.mixcr.basictypes.Clone
import com.milaboratory.mixcr.cli.MultipleMetricsInOneFile
import com.milaboratory.mixcr.postanalysis.overlap.OverlapType
import com.milaboratory.mixcr.postanalysis.plots.ColorKey
import com.milaboratory.mixcr.postanalysis.plots.HeatmapParameters
import com.milaboratory.mixcr.postanalysis.plots.Overlap.dataFrame
import com.milaboratory.mixcr.postanalysis.plots.Overlap.filterOverlap
import com.milaboratory.mixcr.postanalysis.plots.Overlap.plots
import com.milaboratory.mixcr.postanalysis.plots.OverlapRow
import com.milaboratory.mixcr.postanalysis.plots.parseFilter
import com.milaboratory.mixcr.postanalysis.ui.PostanalysisParametersOverlap
import org.jetbrains.kotlinx.dataframe.DataFrame
import picocli.CommandLine.Command
import picocli.CommandLine.Option

@Command(description = ["Export overlap heatmaps"])
class CommandPaExportPlotsOverlap : MultipleMetricsInOneFile, CommandPaExportPlotsHeatmapWithGroupBy() {
    @Option(description = ["Don't add dendrograms"], names = ["--no-dendro"])
    var noDendro = false

    @Option(
        description = ["Add color key layer to the heatmap. One may write `--color-key x_meta` to draw color key horizontally (default) or `--color-key y_meta` to draw vertically."],
        names = ["--color-key"],
        paramLabel = "<meta>"
    )
    var colorKeysParam: List<String> = mutableListOf()

    @Option(description = ["Fill diagonal line"], names = ["--fill-diagonal"])
    var fillDiagonal = false

    @Option(
        description = [
            "Select specific metrics to export.",
            "Possible values are: \${COMPLETION-CANDIDATES}"
        ],
        names = ["--metric"],
        paramLabel = "<metric>"
    )
    var metrics: List<OverlapType>? = null

    override fun validate() {
        super.validate()
        validateNonPdf(out, metrics)
    }

    private fun DataFrame<OverlapRow>.filterOverlapByMetadata(): DataFrame<OverlapRow> {
        var result = this
        filterByMetadata?.let { filterByMetadata ->
            for (f in filterByMetadata) {
                val filter = metadataDf!!.parseFilter(f)
                result = filterOverlap(filter, result)
            }
        }
        return result
    }

    override fun run(result: PaResultByGroup) {
        val ch = result.schema.getGroup<Clone>(PostanalysisParametersOverlap.Overlap)
        val df: DataFrame<OverlapRow> = dataFrame(
            result.result.forGroup(ch),
            metrics,
            fillDiagonal,
            metadataDf
        ).filterOverlapByMetadata()
        if (df.rowsCount() == 0) return
        if (df["weight"].distinct().toList().size <= 1) return
        val colorKeys = colorKeysParam.map { key ->
            when {
                key.startsWith("x_") -> ColorKey(key, Bottom)
                key.startsWith("y_") -> ColorKey(key, Left)
                else -> ColorKey("x_$key".lowercase(), Bottom)
            }
        }
        val par = HeatmapParameters(
            !noDendro,
            !noDendro,
            colorKeys,
            groupBy,
            hLabelsSize,
            vLabelsSize,
            false,
            parsePalette(),
            width,
            height
        )
        val plots = plots(df, par)
        writePlots(result.group, plots)
    }
}
